.github/workflows/php.yml

```diff
@@ -10,28 +10,45 @@ jobs:
   build:
 
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        php-versions: ['7.1', '7.4', '8.0', '8.1']
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
+
+    - name: Set up PHP
+      uses: shivammathur/setup-php@v2
+      with:
+        php-version: ${{ matrix.php-versions }}
+        tools: composer, phpunit
+        extensions: intl, json, mbstring, xdebug
+        coverage: xdebug
+      env:
+        COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
     - name: Validate composer.json and composer.lock
       run: composer validate
 
     - name: Cache Composer packages
       id: composer-cache
-      uses: actions/cache@v2
+      uses: actions/cache@v3
       with:
         path: vendor
-        key: ${{ runner.os }}-php-${{ hashFiles('**/composer.lock') }}
+        key: ${{ runner.os }}-php-${{ hashFiles('**/composer.json') }}
         restore-keys: |
           ${{ runner.os }}-php-
 
     - name: Install dependencies
       if: steps.composer-cache.outputs.cache-hit != 'true'
-      run: composer install --prefer-dist --no-progress --no-suggest
+      run: composer update --prefer-dist --no-progress --no-suggest
```

Comment:
Are you sure about `composer update` in this workflow. Doesn't that create a moving target? Doesn't composer install work with the lock file?

Reply:
You don't have a lockfile committed :) General practice in CI is that for "projects" (a complete app with many dependencies) you would commit the lockfile and use `composer install`, but for "libraries" (a set of features intended to be used in a project) you would exclude the lockfile and use `composer update`. There are exceptions in both directions, but since you haven't been using a lockfile in this repo I think that is in line with standard library practice and we should use `composer update`.

Reply:
fair enough